### ISSUE TITLE
[suggestion] New diagnostic: Use AllItemsConstraint for better assertion messages in case of failure

### ISSUE THREAD

**Issue by Bartleby2718** [nunit/nunit.analyzers#748]
Today, NUnit.Analyzers (4.2.0) doesn't flag the following code:
```cs
var numbers = new[] { 1, 2, 3 };
Assert.That(numbers.All<int>(x => x % 2 == 0));
```
However, NUnit has [a constraint to support this use case](https://docs.nunit.org/api/NUnit.Framework.Constraints.ConstraintExpression.html#NUnit_Framework_Constraints_ConstraintExpression_Matches__1_System_Predicate___0__):
```cs
Assert.That(numbers, Has.All.Matches<int>(x => x % 2 == 0));
```
so I think `NUnit.Analyzers` should suggest a fix for a better error message.

Notes:
1. It appears that the type parameter is required for `Matches`, but I'm not sure if Roslyn will be able to deduce it if the type parameter is absent (i.e.  convert `Assert.That(numbers.All(x => x % 2 == 0));` to the same `Assert.That(numbers, Has.All.Matches<int>(x => x % 2 == 0));`).
2. There's a typo in the link (`custom custom`), but I'm not sure where it's source-controlled.

**Comment by Bartleby2718** [comment]
@manfred-brands I see that `Is` and `Has` are basically the same:
- https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Is.cs#L28-L33
- https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Has.cs#L27-L32

Questions
1. Is there another one that does the same?
2. Which one should we choose?
